### PR TITLE
add revisions garbage collection mechanism

### PR DESCRIPTION
**Pull request by erdun**
1. add revisions garbage collection mechanism, default revision limit is 50.
2. revision in use won't be deleted

### CODE REVIEW

File: pkg/controller/v1alpha2/applicationconfiguration/component.go [hunk 1]
```diff
@@ -169,9 +179,74 @@ func (c *ComponentHandler) createControllerRevision(mt metav1.Object, obj runtim
...
+				liveHashes[component.RevisionName] = true
+			}
+		}
+	}
+
+	toKeep := c.RevisionLimit + len(liveHashes)
+	toKill := len(revisions.Items) - toKeep
+	if toKill <= 0 {
+		return nil
+	}
```
Comment (wonderflow) [inline pkg/controller/v1alpha2/applicationconfiguration/component.go:229]: Could you put it into one separated Function and add a unit test for it

File: pkg/controller/v1alpha2/applicationconfiguration/component.go [hunk 2]
```diff
@@ -169,9 +179,74 @@ func (c *ComponentHandler) createControllerRevision(mt metav1.Object, obj runtim
...
+		}
+		// Clean up
+		revisionToClean := revision
+		if err := c.Client.Delete(context.TODO(), &revisionToClean); err != nil {
+			return err
+		}
+		c.Logger.Info(fmt.Sprintf("ControllerRevision %s deleted", revision.Name))
+		toKill--
+	}
+	return nil
```
Comment (wonderflow) [inline pkg/controller/v1alpha2/applicationconfiguration/component.go:247]: Could you split it into two parts(functions), one part is sort and another part is real delete. 

And for delete, we should add a TODO for auditing in the future

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+// Args args used by controller
+type Args struct {
+	// RevisionLimit is the maximum number of revisions that will be maintained.
+	// The default value is 10.
```
Comment (wonderflow) [inline pkg/controller/controller.go:19]: Make the default number be 50 will be better I think.

File: pkg/controller/v1alpha2/applicationconfiguration/component.go (hunk 1 above)
Comment (erdun) [reply pkg/controller/v1alpha2/applicationconfiguration/component.go:229]: 👌

File: pkg/controller/v1alpha2/applicationconfiguration/component.go (hunk 2 above)
Comment (erdun) [reply pkg/controller/v1alpha2/applicationconfiguration/component.go:247]: 👌

File: pkg/controller/controller.go (hunk 3 above)
Comment (erdun) [reply pkg/controller/controller.go:19]: 👌

**Review by wonderflow — APPROVED** [review]
LGTM, ping @ryanzhang-oss

**Comment by Diddaa** [comment]
LGTM

File: pkg/controller/v1alpha2/applicationconfiguration/component.go [hunk 4]
```diff
@@ -170,9 +179,86 @@ func (c *ComponentHandler) createControllerRevision(mt metav1.Object, obj runtim
...
+	// get all revisions used and skipped
+	for _, appConfig := range appConfigs {
+		for _, component := range appConfig.Spec.Components {
+			if component.RevisionName != "" {
+				liveHashes[component.RevisionName] = true
+			}
+		}
+	}
+
+	toKeep := revisionLimit + len(liveHashes)
```
Comment (ryanzhang-oss) [inline pkg/controller/v1alpha2/applicationconfiguration/component.go:205]: where is toKeep used other than immediately below?

File: pkg/controller/v1alpha2/applicationconfiguration/component.go [hunk 5]
```diff
@@ -170,9 +179,86 @@ func (c *ComponentHandler) createControllerRevision(mt metav1.Object, obj runtim
...
+	controllerRevisions, toKill, liveHashes := sortedControllerRevision(appConfigs.Items, revisions.Items, c.RevisionLimit)
+	for _, revision := range controllerRevisions {
+		if toKill <= 0 {
+			break
+		}
+		if hash := revision.GetName(); liveHashes[hash] {
+			continue
+		}
+		// Clean up
+		revisionToClean := revision
```
Comment (ryanzhang-oss) [inline pkg/controller/v1alpha2/applicationconfiguration/component.go:252]: same, why creating another variable?

**Pull request merged**